src/Client/ClientBaseHelpers.cpp

```diff
@@ -237,4 +241,53 @@ void highlight(const String & query, std::vector<replxx::Replxx::Color> & colors
 }
 #endif
 
+String formatQuery(String query)
+{
+    const unsigned max_parser_depth = DBMS_DEFAULT_MAX_PARSER_DEPTH;
+    const unsigned max_parser_backtracks = DBMS_DEFAULT_MAX_PARSER_BACKTRACKS;
+
+    ParserQuery parser(query.data() + query.size(), /*allow_settings_after_format_in_insert_=*/ false, /*implicit_select_=*/ false);
+
+    String res;
+    res.reserve(query.size());
+
+    const char * begin = query.data();
+    const char * pos = begin;
+    const char * end = begin + query.size();
+    while (pos < end)
+    {
+        const char * query_start = pos;
+
+        Tokens tokens(pos, end);
+        IParser::Pos token_iterator(tokens, max_parser_depth, max_parser_backtracks);
+        while (token_iterator->type != TokenType::Semicolon && token_iterator.isValid())
+            ++token_iterator;
+        const char * query_end = token_iterator->end;
+        bool has_semicolon = token_iterator.isValid() && token_iterator->type == TokenType::Semicolon;
```

Comment:
Why do you need this? 

I think this approach https://github.com/ClickHouse/ClickHouse/blob/6b938d60f82fe2ff3f73b16cff8254fefe24eef1/src/Common/Scheduler/Workload/WorkloadEntityKeeperStorage.cpp#L248 will be suitable.

Reply:
It was needed to detect does query has semicolon or not, but it can be simplified. Done.